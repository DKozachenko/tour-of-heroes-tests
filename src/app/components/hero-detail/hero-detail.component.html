<div *ngIf="hero" pw-automation-id="wrapper">
  <h2 automation-id="heading" pw-automation-id="heading">
    {{ hero.name | uppercase }} Details
  </h2>
  <div automation-id="hero-id" pw-automation-id="hero-id">
    <span>id: </span>{{ hero.id }}
  </div>
  <div>
    <label
      for="hero-name"
      automation-id="hero-name-label"
      pw-automation-id="hero-name-label"
      >Hero name:
    </label>
    <input
      id="hero-name"
      [(ngModel)]="hero.name"
      placeholder="Hero name"
      automation-id="hero-name-input"
      pw-automation-id="hero-name-input"
    />
  </div>
  <button
    type="button"
    (click)="goBack()"
    automation-id="back-button"
    pw-automation-id="back-button"
  >
    go back
  </button>
  <button
    type="button"
    (click)="save()"
    automation-id="save-button"
    pw-automation-id="save-button"
  >
    save
  </button>
</div>
