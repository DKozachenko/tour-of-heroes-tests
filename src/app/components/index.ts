export { DashboardComponent } from './dashboard/dashboard.component';
export { HeroDetailComponent } from './hero-detail/hero-detail.component';
export { HeroSearchComponent } from './hero-search/hero-search.component';
export { HeroesComponent } from './heroes/heroes.component';
export { MessagesComponent } from './messages/messages.component';
export { AppComponent } from './app/app.component';
