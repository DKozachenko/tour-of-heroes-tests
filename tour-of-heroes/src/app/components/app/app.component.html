<h1 automation-id="heading">Tour of Heroes</h1>
<nav automation-id="navigation">
  <a routerLink="/dashboard" automation-id="navigation-link">Dashboard</a>
  <a routerLink="/heroes" automation-id="navigation-link">Heroes</a>
</nav>
<router-outlet></router-outlet>
<app-messages automation-id="messages"></app-messages>
