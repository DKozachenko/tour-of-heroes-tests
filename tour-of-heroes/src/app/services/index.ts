export { HeroService } from './hero/hero.service';
export {
  InMemoryDataService,
  HEROES,
} from './in-memory-data/in-memory-data.service';
export { MessageService } from './message/message.service';
