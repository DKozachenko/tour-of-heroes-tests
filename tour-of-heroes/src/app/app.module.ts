import { BrowserModule } from '@angular/platform-browser';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { AppComponent } from './app.component';
import { AppRoutingModule } from './app-routing.module';
import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
import { InMemoryDataService } from './services';

import {
  DashboardComponent,
  HeroDetailComponent,
  HeroesComponent,
  HeroSearchComponent,
  MessagesComponent,
} from './components';

@NgModule({
  imports: [
    BrowserModule,
    FormsModule,
    AppRoutingModule,
    HttpClientModule,

    // The HttpClientInMemoryWebApiModule module intercepts HTTP requests
    // and returns simulated server responses.
    // Remove it when a real server is ready to receive requests.
    HttpClientInMemoryWebApiModule.forRoot(InMemoryDataService, {
      dataEncapsulation: false,
    }),
  ],
  declarations: [
    AppComponent,
    DashboardComponent,
    HeroesComponent,
    HeroDetailComponent,
    MessagesComponent,
    HeroSearchComponent,
  ],
  bootstrap: [AppComponent],
})
export class AppModule {}
