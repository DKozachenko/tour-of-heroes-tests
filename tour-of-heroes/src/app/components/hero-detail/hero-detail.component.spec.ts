import { DebugElement } from '@angular/core';
import { NgModel } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { ActivatedRoute } from '@angular/router';
import { Location } from '@angular/common';
import { MockBuilder, MockRender, MockedComponentFixture } from 'ng-mocks';
import { of } from 'rxjs';
import { mock, instance, when, verify, anyNumber, anything } from 'ts-mockito';
import { HeroService, HEROES } from '../../services';
import { Hero } from '../../models';
import { HeroDetailComponent } from './hero-detail.component';
import { AppModule } from '../../../app/app.module';

class PageObject {
  private fixtureDebugElement: DebugElement;

  private getElementByAutomationId(id: string): DebugElement {
    return this.fixtureDebugElement.query(By.css(`[automation-id=${id}]`));
  }

  constructor(fixture: MockedComponentFixture<HeroDetailComponent>) {
    this.fixtureDebugElement = fixture.debugElement;
  }

  get heading(): DebugElement {
    return this.getElementByAutomationId('heading');
  }

  get heroId(): DebugElement {
    return this.getElementByAutomationId('hero-id');
  }

  get heroNameLabel(): DebugElement {
    return this.getElementByAutomationId('hero-name-label');
  }

  get heroNameInput(): DebugElement {
    return this.getElementByAutomationId('hero-name-input');
  }

  get backButton(): DebugElement {
    return this.getElementByAutomationId('back-button');
  }

  get saveButton(): DebugElement {
    return this.getElementByAutomationId('save-button');
  }
}

describe('HeroDetailComponent', () => {
  let mockHeroService: HeroService;
  let mockLocation: Location;
  let mockRoute: ActivatedRoute;

  beforeEach(() => {
    mockHeroService = mock(HeroService);
    mockLocation = mock(Location);
    mockRoute = mock(ActivatedRoute);

    return MockBuilder(HeroDetailComponent, AppModule)
      .mock(HeroService, instance(mockHeroService))
      .mock(Location, instance(mockLocation))
      .mock(ActivatedRoute, instance(mockRoute));
  });

  function createFixture(): MockedComponentFixture<HeroDetailComponent> {
    return MockRender(HeroDetailComponent);
  }

  function createComponent(): HeroDetailComponent {
    return createFixture().point.componentInstance;
  }

  function mockCalls(): void {
    const mockId = anyNumber();
    const mockHero = HEROES[0];

    when(mockRoute.snapshot).thenReturn({
      paramMap: new Map([['id', mockId.toString()]]),
    } as any);
    when(mockHeroService.getHero(mockId)).thenReturn(of(mockHero));
  }

  describe('Functional', () => {
    describe('getHero', () => {
      it('should call "getHero" service method while creating', () => {
        const mockId = anyNumber();
        const mockHero = HEROES[0];

        when(mockRoute.snapshot).thenReturn({
          paramMap: new Map([['id', mockId.toString()]]),
        } as any);
        when(mockHeroService.getHero(mockId)).thenReturn(of(mockHero));

        createComponent();
        verify(mockHeroService.getHero(mockId)).once();
      });

      it('should call "getHero" service method via "getHeroes" component method', () => {
        const mockId = anyNumber();
        const mockHero = HEROES[0];

        when(mockRoute.snapshot).thenReturn({
          paramMap: new Map([['id', mockId.toString()]]),
        } as any);
        when(mockHeroService.getHero(mockId)).thenReturn(of(mockHero));

        const component = createComponent();
        component.getHero();
        verify(mockHeroService.getHero(mockId)).twice();
      });
    });

    it('should set "hero" property as existed hero if id exists', () => {
      const existedId = 12;
      const existedHero = <Hero>HEROES.find((hero) => hero.id === existedId);

      when(mockRoute.snapshot).thenReturn({
        paramMap: new Map([['id', existedId.toString()]]),
      } as any);
      when(mockHeroService.getHero(existedId)).thenReturn(of(existedHero));

      const component = createComponent();
      verify(mockHeroService.getHero(existedId)).once();
      expect(component.hero).toEqual(existedHero);
    });

    it('should call "back" Location method via "goBack" component method', () => {
      mockCalls();

      const component = createComponent();
      component.goBack();
      verify(mockLocation.back()).once();
    });

    describe('updateHero', () => {
      it('should not call "updateHero" service method via "save" component method if "hero" property is not set', () => {
        mockCalls();
        const component = createComponent();
        component.hero = undefined;
        component.save();
        verify(mockHeroService.updateHero(anything())).never();
      });

      it('should call "updateHero" service method and "goBack" component method via "save" component method if "hero" property is set', () => {
        mockCalls();
        const mockHero = HEROES[0];
        when(mockHeroService.updateHero(mockHero)).thenReturn(of(undefined));

        const component = createComponent();
        component.hero = mockHero;
        const spyOnGoBack = jest.spyOn(component, 'goBack');
        component.save();
        verify(mockHeroService.updateHero(mockHero)).once();
        expect(spyOnGoBack).toHaveBeenCalled();
      });
    });
  });

  describe('Layout', () => {
    it('should not contain html if "hero" property is not set', () => {
      mockCalls();
      const fixture = createFixture();
      const component = fixture.point.componentInstance;
      component.hero = undefined;

      const fixtureDebugElement = fixture.debugElement;
      expect(fixtureDebugElement.nativeElement.innerHtml).toBeUndefined();
      expect(fixtureDebugElement.nativeElement).toMatchSnapshot();
    });

    it('should contain hero description and action buttons if "hero" property is set', () => {
      mockCalls();
      const fixture = createFixture();
      const component = fixture.point.componentInstance;
      const mockHero = HEROES[0];
      component.hero = mockHero;

      const pageObject = new PageObject(fixture);
      expect(pageObject.heading.nativeElement.textContent).toBe(
        `${mockHero.name.toUpperCase()} Details`
      );
      expect(pageObject.heading.nativeElement).toMatchSnapshot();
      expect(pageObject.heroId.nativeElement.textContent).toBe(
        `id: ${mockHero.id}`
      );
      expect(pageObject.heroId.nativeElement).toMatchSnapshot();
      expect(pageObject.heroNameLabel.nativeElement.textContent).toBe(
        'Hero name: '
      );
      expect(pageObject.heroNameLabel.nativeElement).toMatchSnapshot();
      expect(pageObject.heroNameInput.injector.get(NgModel).model).toBe(
        mockHero.name
      );
      expect(pageObject.heroNameInput.nativeElement).toMatchSnapshot();
      expect(pageObject.backButton).not.toBeNull();
      expect(pageObject.backButton.nativeElement).toMatchSnapshot();
      expect(pageObject.saveButton).not.toBeNull();
      expect(pageObject.saveButton.nativeElement).toMatchSnapshot();
    });

    it('should call "goBack" component method if go back button has clicked', () => {
      mockCalls();
      const fixture = createFixture();
      const component = fixture.point.componentInstance;

      const pageObject = new PageObject(fixture);
      const spyOnGoBack = jest.spyOn(component, 'goBack');
      pageObject.backButton.triggerEventHandler('click');
      expect(spyOnGoBack).toHaveBeenCalled();
    });

    it('should call "save" component method if save button has clicked', () => {
      const mockId = anyNumber();
      const mockHero = HEROES[0];
      when(mockRoute.snapshot).thenReturn({
        paramMap: new Map([['id', mockId.toString()]]),
      } as any);
      when(mockHeroService.getHero(mockId)).thenReturn(of(mockHero));
      when(mockHeroService.updateHero(anything())).thenReturn(of(anything()));

      const fixture = createFixture();
      const component = fixture.point.componentInstance;

      const pageObject = new PageObject(fixture);
      const spyOnSave = jest.spyOn(component, 'save');
      pageObject.saveButton.triggerEventHandler('click');
      expect(spyOnSave).toHaveBeenCalled();
    });
  });
});
