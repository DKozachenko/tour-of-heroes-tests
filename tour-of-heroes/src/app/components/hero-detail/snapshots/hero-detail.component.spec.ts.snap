// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroDetailComponent Layout should contain hero description and action buttons if "hero" property is set 1`] = `
<h2
  automation-id="heading"
>
  DR. NICE Details
</h2>
`;

exports[`HeroDetailComponent Layout should contain hero description and action buttons if "hero" property is set 2`] = `
<div
  automation-id="hero-id"
>
  <span>
    id: 
  </span>
  12
</div>
`;

exports[`HeroDetailComponent Layout should contain hero description and action buttons if "hero" property is set 3`] = `
<label
  automation-id="hero-name-label"
  for="hero-name"
>
  Hero name: 
</label>
`;

exports[`HeroDetailComponent Layout should contain hero description and action buttons if "hero" property is set 4`] = `
<input
  automation-id="hero-name-input"
  id="hero-name"
  placeholder="Hero name"
  pw-automation-id="hero-name-input"
/>
`;

exports[`HeroDetailComponent Layout should contain hero description and action buttons if "hero" property is set 5`] = `
<button
  automation-id="back-button"
  type="button"
>
   go back 
</button>
`;

exports[`HeroDetailComponent Layout should contain hero description and action buttons if "hero" property is set 6`] = `
<button
  automation-id="save-button"
  pw-automation-id="save-button"
  type="button"
>
   save 
</button>
`;

exports[`HeroDetailComponent Layout should not contain html if "hero" property is not set 1`] = `
<div
  id="root6"
>
  <app-hero-detail>
    <div
      pw-automation-id="wrapper"
    >
      <h2
        automation-id="heading"
      >
        DR. NICE Details
      </h2>
      <div
        automation-id="hero-id"
      >
        <span>
          id: 
        </span>
        12
      </div>
      <div>
        <label
          automation-id="hero-name-label"
          for="hero-name"
        >
          Hero name: 
        </label>
        <input
          automation-id="hero-name-input"
          id="hero-name"
          placeholder="Hero name"
          pw-automation-id="hero-name-input"
        />
        
      </div>
      <button
        automation-id="back-button"
        type="button"
      >
         go back 
      </button>
      <button
        automation-id="save-button"
        pw-automation-id="save-button"
        type="button"
      >
         save 
      </button>
    </div>
    
  </app-hero-detail>
</div>
`;
